jupyterlab/upgrade_extension.py

```diff
@@ -40,40 +77,68 @@ def update_extension(target, branch=DEFAULT_COOKIECUTTER_BRANCH, interactive=Tru
     # Pull in the relevant config
     # Pull in the Python parts if possible
     # Pull in the scripts if possible
-    target = osp.abspath(target)
-    package_file = osp.join(target, "package.json")
-    setup_file = osp.join(target, "setup.py")
-    if not osp.exists(package_file):
-        raise RuntimeError("No package.json exists in %s" % target)
+    target = Path(target).resolve()
+    package_file = target / "package.json"
+    pyproject_file = target / "pyproject.toml"
+    setup_file = target / "setup.py"
+    if not package_file.exists():
+        msg = f"No package.json exists in {target!s}"
+        raise RuntimeError(msg)
 
     # Infer the options from the current directory
     with open(package_file) as fid:
         data = json.load(fid)
 
-    if osp.exists(setup_file):
-        python_name = (
-            subprocess.check_output([sys.executable, "setup.py", "--name"], cwd=target)
-            .decode("utf8")
-            .strip()
-        )
-    else:
-        python_name = data["name"]
-        if "@" in python_name:
-            python_name = python_name[1:].replace("/", "_").replace("-", "_")
+    python_name = None
+    if pyproject_file.exists():
+        pyproject = tomllib.loads(pyproject_file.read_text())
+        python_name = pyproject.get("project", {}).get("name")
+
+    if python_name is None:
+        if setup_file.exists():
+            python_name = (
+                subprocess.check_output([sys.executable, "setup.py", "--name"], cwd=target)
+                .decode("utf8")
+                .strip()
+            )
+        else:
+            python_name = data["name"]
+            if "@" in python_name:
+                python_name = python_name[1:].replace("/", "_").replace("-", "_")
 
-    output_dir = osp.join(target, "_temp_extension")
-    if osp.exists(output_dir):
+    output_dir = target / "_temp_extension"
+    if output_dir.exists():
         shutil.rmtree(output_dir)
 
     # Build up the cookiecutter args and run the cookiecutter
+    author = data.get("author", "<author_name>")
+    author_email = "<author_email>"
+    if isinstance(author, dict):
+        author = author.get("name", "<author_name>")
+        author_email = author.get("email", author_email)
```

Comment:
```suggestion
        author_email = author.get("email", author_email)
        author = author.get("name", "<author_name>")
```

Or even better, switch to explicit `author_name`. Currently we get:

```python
AttributeError: 'str' object has no attribute 'get'
```

(as of now `author` gets overridden with a string).

Reply:
Good catch